### PR TITLE
Suport for custom repo providers

### PR DESCRIPTION
**Pull request by g-braeunlich**
This MR adds support for custom repo providers via the configurable `repo_provider` field in `PersistentBinderSpawner`.
At the same time it also deduplicates logic which was present in both UI and the spawner.

### CODE REVIEW

**Comment by MridulS** [comment]
Thanks for this @g-braeunlich! :)

File: persistent_binderhub/files/jupyterhub/persistent_bhub_config.py [hunk 1]
```diff
@@ -13,6 +13,15 @@
 from jupyterhub.apihandlers.users import admin_or_self
 from jupyterhub.apihandlers.base import APIHandler
 from kubespawner import KubeSpawner
+from traitlets import List
+
+DEFAULT_PROVIDERS = (
+    {"prefix": "gh", "name": "GitHub", "hostname": "github.com"},
+    {"prefix": "git", "name": "Git", "hostname": None},
+    {"prefix": "gl", "name": "GitLab", "hostname": "gitlab.com"},
+)
```
Comment (MridulS) [inline persistent_binderhub/files/jupyterhub/persistent_bhub_config.py:23]: Could you also add `Gist` here?
```
{"prefix": "gist", "name": "Gist", "hostname": "gist.github.com"},
```

Comment (g-braeunlich) [reply persistent_binderhub/files/jupyterhub/persistent_bhub_config.py:23]: done. force pushed into the previous commit

**Comment by MridulS** [comment]
thanks!

**Comment by g-braeunlich** [comment]
You are welcome!

**Pull request merged**